_drafts/unit-test-mocking.md

```diff
@@ -0,0 +1,1006 @@
+CppUMock. First, it requires a good bit of boiler plate which leads to verbose
+unit testing code. Where other libraries use macros to make things simpler and
+more concise, CppUMock requires you to write everything by hand. And second, the
+documentation leaves a lot to be desired, and I commonly find myself looking at
+the source code for extra functionality that isn't documented.
+
+### Fake Function Framework (fff)
+
+[Fake Function Framework (fff)](https://github.com/meekrosoft/fff) is the
+simplest way to get started with mocking. It is a header-only library which
+makes it trivial to include in your project.
+
+Below, we implement the same functionality as the example from the CppUMock
+section.
+
+```c
+TEST(TestAllocateBuffer, Test_AllocateBufferBasic) {
+  // Set return value
+  MOCK_my_malloc.return_value = NULL;
+
+  uint8_t *buf = allocate_buffer(32);
+
+  // Validate expectations
+  LONGS_EQUAL(1, MOCK_my_malloc.call_count);
+  LONGS_EQUAL(MOCK_my_malloc.arg0_val, 32);
+  POINTERS_EQUAL(NULL, buf);
+
+  // Get a real buffer we can write into
+  void *real_buf = malloc(32);
+  MOCK_my_malloc.return_value = real_buf;
+  uint8_t *buf = allocate_buffer(32);
+  CHECK(buf != NULL);
+}
+```
+
+### CMock
+
+CMock is a mocking library from the group behind
+[ThrowTheSwitch.org](http://www.throwtheswitch.org/), which also makes the Unity
+unit testing framework[^unity]. It is compatible with C and includes some
+tooling around running and generating boilerplate for you given a set of header
+files. The flow is similar to CppUMock in that you define your expectations
+first and then call the implementation under test.
+
+```c
+TEST(TestAllocateBuffer, Test_AllocateBufferBasic) {
+  // Set return value and expectation
+  my_malloc_ExpectAndReturn(32, NULL);
+
+  uint8_t *buf = allocate_buffer(32);
+
+  // Validate expectations
+  POINTERS_EQUAL(NULL, buf);
+
+  // Get a real buffer we can write into
+  void *real_buf = malloc(32);
+  my_malloc_ExpectAndReturn(32, real_buf);
+  uint8_t *buf = allocate_buffer(32);
+  CHECK(buf != NULL);
+}
+```
+
+The main reason I do not use the tools from ThrowTheSwitch is that they require
+Ruby, and I haven't seen any significant number of embedded engineers or
+projects in the wild using Ruby. With this in mind, I prefer to use other
+frameworks and tools so that my team and I don't need to manage a Ruby
+environment (alongside the usual GCC and Python environments).
+
+## Real World Unit Test Example
+
+Throughout this post, we are going to build a way for a remote system (such as
+an iOS app) connected to the device over some transport (e.g. BLE or serial) to
+be able to read and write key/value pairs to and from the device's internal file
+system.
+
+This is likely a system that most embedded engineers have either built or worked
+with.
+
+Below is a diagram of each piece in the system mentioned above.
+
+![](/img/unit-testing-mocking/diagram.png)
```

Comment:
Might call out that this diagram is our unit test stack (i.e "Fake NOR Flash" instead of just "NOR Flash") ... the previous paragraph had me thinking this was the actual system diagram